api/src/main/java/edu/cornell/mannlib/vitro/webapp/audit/storage/AuditDAOJena.java

```diff
@@ -0,0 +1,500 @@
+                                auditModel.createResource(graphName));
+                    }
+
+                    // Record graph of added statements
+                    if (!isEmpty(addedModel)) {
+                        String addedName = graphUri + "/added";
+                        Model addedAuditModel = auditStore.getNamedModel(addedName);
+                        if (addedAuditModel != null) {
+                            addedAuditModel.add(addedModel);
+                            graphResource.addProperty(auditModel.createProperty(AuditVocabulary.PROP_GRAPH_ADDED),
+                                    auditModel.createResource(addedName));
+                        }
+                    }
+
+                    // Record graph of removed statements
+                    if (!isEmpty((removedModel))) {
+                        String removedName = graphUri + "/removed";
+                        Model removedAuditModel = auditStore.getNamedModel(removedName);
+                        if (removedAuditModel != null) {
+                            removedAuditModel.add(removedModel);
+                            graphResource.addProperty(auditModel.createProperty(AuditVocabulary.PROP_GRAPH_REMOVED),
+                                    auditModel.createResource(removedName));
+                        }
+                    }
+
+                    // Increment to create a new URI for any additional graphs
+                    graphCount++;
+                }
+            }
+
+            // Commit the changes to the audit store
+            auditStore.commit();
+        } finally {
+            auditStore.end();
+        }
+    }
+
+    @Override
+    public AuditResults find(long offset, int limit, long startDate, long endDate, String userUri, String graphUri,
+            boolean order) {
+        long total = 0;
+        List<AuditChangeSet> datasets = new ArrayList<AuditChangeSet>();
+
+        // Get the audit dataset
+        Dataset auditStore = getDataset();
+        if (auditStore == null) {
+            return null;
+        }
+
+        // Indicate that we are reading from the audit store
+        auditStore.begin(ReadWrite.READ);
+        try {
+            StringBuilder queryString;
+            Query query;
+            QueryExecution qexec;
+
+            // SPARQL query to retrieve overall change sets, in reverse chronological order, with pagination
+            queryString = new StringBuilder();
+            queryString.append("SELECT ?dataset");
+            queryString.append(" WHERE {");
+            queryString.append("   ?dataset a <").append(AuditVocabulary.TYPE_CHANGESET).append("> . ");
+            queryString.append("   ?dataset <").append(AuditVocabulary.PROP_DATE).append("> ?date . ");
+            if (!StringUtils.isBlank(userUri)) {
+                queryString.append("   ?dataset <").append(AuditVocabulary.PROP_USER).append("> <").append(userUri)
```

Comment:
This part of the code is vulnerable to SparQL Injection.
For example I could send a filter param for userUri like this
`http://vivoweb.org/audit/resource/unknown> %7d  LIMIT 999999 #`

It might be possible to request any data in db through this. But since this is admin only feature I guess it's not a problem at this moment

Reply:
Thanks for catching that. I think I made it safer now. Please check again.

Reply:
Injection resolved. Great job!